rust/dialog-common-macros/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+//! must be defined in their own crate.
+
+use proc_macro::TokenStream;
+mod provider;
+mod test;
+
+// disabling because we don't want to add crate dependencies just for this
+#[cfg(not(doctest))]
+/// A cross-platform test macro with automatic service provisioning.
+///
+/// # CI Test Matrix
+///
+/// The macro generates code that supports these CI configurations:
+///
+/// 1. `cargo test` - All tests run natively (unit tests + integration tests inline)
+/// 2. `cargo test --target wasm32-unknown-unknown` - Unit tests run in wasm
+/// 3. `RUSTFLAGS="--cfg dialog_test_wasm" cargo test` - Integration tests run in wasm
+///    (native provider spawns wasm inner tests)
+///
+/// # Usage
+///
+/// ## Unit tests
+///
+/// Tests that do not require external services:
+///
+/// ```rs
+/// // Sync test
+/// #[dialog_common::test]
+/// fn it_works() {
+///     assert_eq!(2 + 2, 4);
+/// }
+///
+/// // Async test
+/// #[dialog_common::test]
+/// async fn it_works_async() -> anyhow::Result<()> {
+///     assert_eq!(2 + 2, 4);
+///     Ok(())
+/// }
+/// ```
+///
+/// Unit tests are gated with `#[cfg(not(dialog_test_wasm))]` so they don't
+/// run during wasm integration test runs (case 3 above).
+///
+/// ## Integration tests
+///
+/// Tests that need external services (S3, databases, servers, etc.):
+///
+/// ### Step 1: Define the address (what the test receives)
+///
+/// The address is serializable so it can be passed to wasm tests via environment:
+///
+/// ```rs
+/// use serde::{Deserialize, Serialize};
+///
+/// #[derive(Debug, Clone, Serialize, Deserialize)]
+/// pub struct ServerAddress {
+///     pub endpoint: String,
+///     pub port: u16,
+/// }
+/// ```
+///
+/// ### Step 2: Define the provider (native only)
+///
+/// Use `#[dialog_common::provider]` to create a function that starts the service
+/// and returns its address:
+///
+/// ```rs
+/// use dialog_common::helpers::{Provider, Service};
+///
+/// pub struct LocalServer { /* ... */ }
+///
+/// impl Provider for LocalServer {
+///     async fn stop(self) -> anyhow::Result<()> {
+///         Ok(())
+///     }
+/// }
+///
+/// #[dialog_common::provider]
+/// async fn server(_settings: ()) -> anyhow::Result<Service<ServerAddress, LocalServer>> {
+///     let server = LocalServer { /* start server */ };
+///     Ok(Service::new(
+///         ServerAddress { endpoint: "...".into(), port: 8080 },
+///         server
+///     ))
+/// }
+/// ```
+///
+/// ### Step 3: Write the test
+///
+/// When a function takes an address parameter, the macro generates:
+///
+/// - **Native inline provider**: Starts service, runs test via `tokio::spawn`, stops service
+/// - **Native wasm-spawn provider**: Starts service, spawns `cargo test --target wasm32...`
+/// - **Wasm inner**: Deserializes address from env var, runs test
+///
+/// ```rs
+/// #[dialog_common::test]
+/// async fn it_connects_to_server(server: ServerAddress) -> anyhow::Result<()> {
```

Comment:
Can you make it explicit in the docs that the parameter is what turns this into an integration test? It took me a a fair amount of squinting to infer that.